### ISSUE TITLE
Look into reducing webpack times

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4484]
This task is automatically imported from the old Task Issue Board and it was originally created by Michał Urbańczyk.
Original issue is [here](https://www.pivotaltracker.com/story/show/182443923).
<hr />

1. See if the "do not copy if not changed" is enough to stop spurious rebuilds.
2. See if esbuild could work as a quick replacement for webpack.

## Comments:
**Michał Urbańczyk** reports a new **STANDUP** for the last Friday (2022-06-10):

**Progress:** Implemented contents check for wasm artifacts, so they are not copied when not actually modified. Observed results, not promising. webpack-dev-server still rebuilds, it just skips reloading the page because it sees no change.
Follow up fixes for the build script after the last deployment. It should be finished by 2022-06-14.

**Next Day:** Next day I will be working on the [##182443923 task](https://www.pivotaltracker.com/story/show/182443923). Try using esbuild. (Enso Bot - Jun 14, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2022-06-13):

**Progress:** Spent some time fixing issues with the build script (regressions from the recent deployment that unfortunately show up only after it).
Started experiments with esbuild. First impressions are good, though I have much to learn. It should be finished by 2022-06-14.

**Next Day:** Next day I will be working on the [##182443923 task](https://www.pivotaltracker.com/story/show/182443923). Try getting `content` package to work with esbuild. (Enso Bot - Jun 14, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2022-06-14):

**Progress:** Further experiments with esbuild. Got the content project to be bundled. Speedup when compared to webpack is around 5-~15 times. I have hoped for more, but this is still a decent improvement.
Encountered issues when I've tried to use watch and serve modes with esbuild. Likely, a custom watcher and server are warranted. As discussed with Wojciech, custom dev server makes much sense in our case.
Esbuild seems nice enough (and faster) to be used by us, but needs more work to actually replace current webpack and its devserver. It should be finished by 2022-06-14.

**Next Day:** Next day I will be working on the [##180302898 task](https://www.pivotaltracker.com/story/show/180302898). Get back to the nightly releases. Look into the firebase auth issue, and take appropriate steps. (Enso Bot - Jun 16, 2022)
<hr />
**Michał Urbańczyk** reports a new **🔴 DELAY** for today (2022-07-07):

**Summary:** There is **31 days** delay in implementation of the [Look into reducing webpack times (#182443923) task](https://www.pivotaltracker.com/story/show/182443923).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** Previously started task was on hold. (Enso Bot - Jul 7, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2022-07-06):

**Progress:** Succeeded in semi-manually serving esbuild-bundled GUI and did rudimentary tests in Chrome. So far looks promising. It should be finished by 2022-07-15.

**Next Day:** Next day I will be working on the [##182443923 task](https://www.pivotaltracker.com/story/show/182443923). Automate the setup. Make ./run ide watch spawn electron-used server with esbuild-bundled files. (No support for dev server yet.) (Enso Bot - Jul 7, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for the last Friday (2022-07-08):

**Progress:** Investigating a new dev-server with live-reload support. Unfortunately, all solutions for esbuild are not adequate for our needs. Discussed the matter with Wojciech, we decided to use a separate live-reload dev-server package. Did initial experiments with `live-server`. Had some issues, because its script injection does not play well with our security policies. Added workaround for now. It should be finished by 2022-07-15.

**Next Day:** Next day I will be working on the [##182443923 task](https://www.pivotaltracker.com/story/show/182443923). Below. (Enso Bot - Jul 11, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2022-07-09):

**Progress:** Built a new watch script based on esbuild and live-server. Observed issue where esbuild watcher is not aware of files copied by the plugins. Eventually managed to create a workaround, but it is clumsy.
The new watch script seems to work in typical scenarios. It should be finished by 2022-07-15.

**Next Day:** Next day I will be working on the [##182443923 task](https://www.pivotaltracker.com/story/show/182443923). Weekly updates and reviewing pivotal. Start work on replacing our current server used in the client with the dev-server. (Enso Bot - Jul 11, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2022-07-11):

**Progress:** Weekly presentation. Going over Pivotal Tasks and discussing them with the management. Refactored workaroundish copy-plugin into a separate package, so it can be reused for packaging the ide client package. Started work on using esbuild in client package. Some cleanups. It should be finished by 2022-07-15.

**Next Day:** Next day I will be working on the [##182443923 task](https://www.pivotaltracker.com/story/show/182443923). Use live-server in the electron client. Finish esbuild usage for client, and remove remaining webpack usage in IDE. (Enso Bot - Jul 12, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2022-07-12):

**Progress:** Implemented esbuild-based deployment for ide client. Replaced ide server with the live-server, as discussed with Wojciech.
Initial tests of generated electron package look reasonably well. It should be finished by 2022-07-15.

**Next Day:** Next day I will be working on the [##182690780 task](https://www.pivotaltracker.com/story/show/182690780). Implement build-script support for generating Java code for dealing with the new parser. (Enso Bot - Jul 14, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2022-07-14):

**Progress:** Discovered serious issues with using live-server from an Electron. Struggled for a while to workaround them but eventually it did not work.
live-server does not properly mark its dependencies for non-dev usage, depends on injected html snippet being accessible through fs — and after jumping all these hoops, it just ended up spewing strange errors from depth of electron that I did not have time to further debug.
Otherwise, applying cleanups and minor fixes. It should be finished by 2022-07-15.

**Next Day:** Next day I will be working on the [##182443923 task](https://www.pivotaltracker.com/story/show/182443923). below (Enso Bot - Jul 15, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for today (2022-07-15):

**Progress:** After discussion with Greg and Sylwia went for a "safe" approach of using the so-far server in the Electron client. Cleanups, fixes, testing.
Fixing docs, wrapping up the PR. It should be finished by 2022-07-15.

**Next Day:** Next day I will be working on the [##182443923 task](https://www.pivotaltracker.com/story/show/182443923). Have a call on August 1st. (Enso Bot - Jul 15, 2022)
<hr />
**Michał Urbańczyk** reports a new **🔴 DELAY** for today (2022-08-09):

**Summary:** There is **25 days** delay in implementation of the [Look into reducing webpack times (#182443923) task](https://www.pivotaltracker.com/story/show/182443923).
It **will cause 28 days delay** for the delivery of this weekly plan.

**Delay Cause:** 1) vacation
2) switching tasks
3) waiting for review / call
4) one additional bug found during testing
all in all there's like 2-days of work-delay IIRC (Enso Bot - Aug 10, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for yesterday (2022-08-08):

**Progress:** Addressing the issue found by Michael, where watch does not work if the WASM build has not already succeeded. Weekly updates. It should be finished by 2022-08-09.

**Next Day:** Next day I will be working on the [##182443923 task](https://www.pivotaltracker.com/story/show/182443923). Complete obtaining feedback and land the PR. (Enso Bot - Aug 10, 2022)
<hr />
**Michał Urbańczyk** reports a new **STANDUP** for today (2022-08-09):

**Progress:** Having confirmed the fix, the PR has landed. It should be finished by 2022-08-09.

**Next Day:** Next day I will be working on the [##182443923 task](https://www.pivotaltracker.com/story/show/182443923). Debug zombie processes on Windows. (Enso Bot - Aug 10, 2022)
<hr />
